pkg/knowledge_base2/properties/resource_property.go

```diff
@@ -0,0 +1,164 @@
+package properties
+
+import (
+	"fmt"
+
+	"github.com/klothoplatform/klotho/pkg/collectionutil"
+	construct "github.com/klothoplatform/klotho/pkg/construct2"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base2"
+)
+
+type (
+	ResourceProperty struct {
+		DefaultValue *construct.ResourceId  `json:"default_value" yaml:"default_value"`
+		AllowedTypes construct.ResourceList `yaml:"allowed_types"`
+		Namespace    bool                   `yaml:"namespace"`
+		SharedPropertyFields
+		*knowledgebase.PropertyDetails
+	}
+)
+
+func (r *ResourceProperty) SetProperty(resource *construct.Resource, value any) error {
+	if val, ok := value.(construct.ResourceId); ok {
+		return resource.SetProperty(r.Path, val)
+	} else if val, ok := value.(construct.PropertyRef); ok {
+		return resource.SetProperty(r.Path, val)
+	}
+	return fmt.Errorf("invalid resource value %v", value)
+}
+
+func (r *ResourceProperty) AppendProperty(resource *construct.Resource, value any) error {
+	return r.SetProperty(resource, value)
+}
+
+func (r *ResourceProperty) RemoveProperty(resource *construct.Resource, value any) error {
+	delete(resource.Properties, r.Path)
+	return nil
+}
+
+func (r *ResourceProperty) Details() *knowledgebase.PropertyDetails {
+	return r.PropertyDetails
+}
+func (r *ResourceProperty) Clone() knowledgebase.Property {
```

Comment:
If `PropertyDetails` isn't a pointer, you don't really need `Clone`, because everything is by-value.
This would clone the object:
```go
var original ResourceProperty
cloned := original
```

Reply:
in the case where theyre pointers though you want to dereference and then do that?

Reply:
yeah, you could
```go
var original *ResourceProperty
cloned := *original
use(&cloned)
```

Reply:
Yeah but what about the case where we have to clone a property without knowing its type? I dont think theres a proper way to do that. Are you proposing still having a clone method, just doing it that way?